client/clientutil/modelinfo.go

```diff
@@ -0,0 +1,580 @@
+// -*- Mode: Go; indent-tabs-mode: t -*-
+
+/*
+ * Copyright (C) 2022 Canonical Ltd
+ *
+ * This program is free software: you can redistribute it and/or modify
+ * it under the terms of the GNU General Public License version 3 as
+ * published by the Free Software Foundation.
+ *
+ * This program is distributed in the hope that it will be useful,
+ * but WITHOUT ANY WARRANTY; without even the implied warranty of
+ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+ * GNU General Public License for more details.
+ *
+ * You should have received a copy of the GNU General Public License
+ * along with this program.  If not, see <http://www.gnu.org/licenses/>.
+ *
+ */
+
+package clientutil
+
+import (
+	"encoding/json"
+	"errors"
+	"fmt"
+	"strings"
+	"text/tabwriter"
+	"time"
+
+	"github.com/snapcore/snapd/asserts"
+	"github.com/snapcore/snapd/i18n"
+	"github.com/snapcore/snapd/strutil"
+	"github.com/snapcore/snapd/timeutil"
+)
+
+var (
+	invalidTypeMessage = i18n.G("invalid type for %q header")
+	errNoMainAssertion = errors.New(i18n.G("device not ready yet (no assertions found)"))
+	errNoSerial        = errors.New(i18n.G("device not registered yet (no serial assertion found)"))
+
+	// this list is a "nice" "human" "readable" "ordering" of headers to print
+	// off, sorted in lexical order with meta headers and primary key headers
+	// removed, and big nasty keys such as device-key-sha3-384 and
+	// device-key at the bottom
+	// it also contains both serial and model assertion headers, but we
+	// follow the same code path for both assertion types and some of the
+	// headers are shared between the two, so it still works out correctly
+	niceOrdering = [...]string{
+		"architecture",
+		"base",
+		"classic",
+		"display-name",
+		"gadget",
+		"kernel",
+		"revision",
+		"store",
+		"system-user-authority",
+		"timestamp",
+		"required-snaps", // for uc16 and uc18 models
+		"snaps",          // for uc20 models
+		"device-key-sha3-384",
+		"device-key",
+	}
+)
+
+type ModelFormatter interface {
+	GetPublisher() string
+	GetEscapedDash() string
+}
+
+// modelAssertJSON is a helper to write out a single assertion in json
+// format that matches how it's done in api_model.go. We have stolen
+// the structure from there.
+type modelAssertJSON struct {
+	Headers map[string]interface{} `json:"headers,omitempty"`
+	Body    string                 `json:"body,omitempty"`
+}
+
+func fmtTime(t time.Time, abs bool) string {
+	if abs {
+		return t.Format(time.RFC3339)
+	}
+	return timeutil.Human(t)
+}
+
+// PrintModelAssertionYAML will format the provided serial or model assertion based on the parameters given in
+// YAML format. The output will be written to the provided io.Writer.
```

Comment:
Can you document the assertion bool? Maybe `printAssertionOnly'? Also, perhaps it makes sense to wrap all the bool flags (maybe more?) in a flags struct (also for the JSON variant)?

Reply:
Yea I like the suggestions to this!

Reply:
Reduced bool by one and moved to struct